zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/callback/ZkAsyncRetryThread.java

```diff
@@ -49,7 +49,9 @@ public void run() {
           context.cancel();
           interrupt();
         } catch (Throwable e) {
-          LOG.error("Error retrying callback " + context, e);
+          LOG.error("Error retrying callback {}, cancelling it", context, e);
+          // Cancel the context so the upstream caller can stop waiting
+          context.cancel();
```

Comment:
Let's cancel in final statement?

Reply:
No, if retry was done successfully, then cancel it prematurely will block the normal logic.

Reply:
As @jiajunwang said, cancel it prematurely in the finally statement will impact the future retries: if this retry fails because of connection loss, it'll continue retrying. If we cancel it in the finally, the operation is marked as done so it won't retry the operation. We only want to cancel it if a non-retryable exception throws.